src/librustc_mir/util/elaborate_drops.rs

```diff
@@ -535,6 +535,114 @@ impl<'l, 'b, 'tcx, D> DropCtxt<'l, 'b, 'tcx, D>
+                 succ: BasicBlock,
+                 index: &Lvalue<'tcx>,
+                 length: &Lvalue<'tcx>,
+                 ety: Ty<'tcx>,
+                 is_cleanup: bool)
+                 -> BasicBlock
+    {
+        let use_ = |lv: &Lvalue<'tcx>| Operand::Consume(lv.clone());
+        let tcx = self.tcx();
+
+        let ref_ty = tcx.mk_ref(tcx.types.re_erased, ty::TypeAndMut {
+            ty: ety,
+            mutbl: hir::Mutability::MutMutable
+        });
+        let ptr = &Lvalue::Local(self.new_temp(ref_ty));
+        let can_go = &Lvalue::Local(self.new_temp(tcx.types.bool));
+
+        let one = self.constant_usize(1);
+        let drop_block = self.elaborator.patch().new_block(BasicBlockData {
+            statements: vec![
+                Statement { source_info: self.source_info, kind: StatementKind::Assign(
+                    ptr.clone(), Rvalue::Ref(
+                        tcx.types.re_erased, BorrowKind::Mut,
+                        self.lvalue.clone().index(use_(index))
+                    ),
+                )},
+                Statement { source_info: self.source_info, kind: StatementKind::Assign(
+                    index.clone(), Rvalue::BinaryOp(BinOp::Add, use_(index), one)
+                )},
+            ],
+            is_cleanup,
+            terminator: Some(Terminator {
+                source_info: self.source_info,
+                kind: TerminatorKind::Resume,
+            })
+        });
+
+        let loop_block = self.elaborator.patch().new_block(BasicBlockData {
+            statements: vec![
+                Statement { source_info: self.source_info, kind: StatementKind::Assign(
+                    can_go.clone(), Rvalue::BinaryOp(BinOp::Lt, use_(index), use_(length))
+                )},
+            ],
+            is_cleanup,
+            terminator: Some(Terminator {
+                source_info: self.source_info,
+                kind: TerminatorKind::if_(tcx, use_(can_go), drop_block, succ)
+            })
+        });
+
+        self.elaborator.patch().patch_terminator(drop_block, TerminatorKind::Drop {
+            location: ptr.clone().deref(),
+            target: loop_block,
+            unwind: unwind
+        });
+
+        loop_block
+    }
+
+    fn open_drop_for_array(&mut self, ety: Ty<'tcx>) -> BasicBlock {
+        debug!("open_drop_for_array({:?})", ety);
+        // FIXME: using an index instead of a pointer to avoid
+        // special-casing ZSTs.
+        let tcx = self.tcx();
+        let index = &Lvalue::Local(self.new_temp(tcx.types.usize));
+        let length = &Lvalue::Local(self.new_temp(tcx.types.usize));
+
+        let unwind = self.unwind.map(|unwind| {
+            self.drop_loop(None, unwind, index, length, ety, true)
+        });
+
+        let is_cleanup = self.is_cleanup;
+        let succ = self.succ; // FIXME(#6393)
+        let loop_block = self.drop_loop(unwind, succ, index, length, ety, is_cleanup);
```

Comment:
`is_cleanup` is technically always false here because this can't be reached from "normal" drop elaboration, but if we fixed drop glue for arrays it would be true.